shell/packages/sandstorm-ui-powerbox/powerbox-server.js

```diff
@@ -184,7 +235,65 @@ specialCaseTypes[Ip.IpInterface.typeId] = function (db, userId, value) {
   } else {
     return [];
   }
-},
+};
+
+specialCaseTypes[Email.EmailVerifier.typeId] = function (db, userId, value) {
+  const results = [];
+
+  results.push({
+    _id: "emailverifier-all",
+    frontendRef: { emailVerifier: {} }
+  });
+
+  for (const name in Accounts.identityServices) {
+    if (Accounts.identityServices[name].isEnabled()) {
+      results.push({
+        _id: "emailverifier-" + name,
+        frontendRef: { emailVerifier: { services: [name] } }
+      });
+    }
+  };
+
+  return results;
+};
+
+specialCaseTypes[Email.VerifiedEmail.typeId] = function (db, userId, value) {
+  const verifierId = value &&
+      Capnp.parse(Email.VerifiedEmail.PowerboxTag, value).verifierId.toString("base64");
+  return getVerifiedEmails(db, userId, verifierId).map(address => ({
+    _id: "email-" + address,
+    frontendRef: { verifiedEmail: { verifierId, address } }
+  }));
+};
+
+function getVerifiedEmails(db, userId, verifierId) {
+  // Get all of the email addresses verified as belonging to the given user using the given
+  // verifier.
+
+  let services = null;
+
+  if (verifierId) {
+    const verifier = db.collections.apiTokens.findOne(verifierId);
+    const verifierInfo = ((verifier || {}).frontendRef || {}).emailVerifier;
+    if (!verifierInfo) return [];  // invalid verifier
+
+    if (verifierInfo.services) {
+      // Limit to the listed services.
+      services = {};
+      verifierInfo.services.forEach(service => services[service] = true);
+    }
+  }
+
+  const user = Meteor.users.findOne(userId);
+  const emails = {};  // map address -> true, for uniquification
+  Meteor.users.find({_id: {$in: SandstormDb.getUserIdentityIds(user)}}).forEach(identity => {
+    if (!services || services[identity.profile.service]) {
+      SandstormDb.getVerifiedEmails(identity).forEach(email => emails[email.email] = true);
```

Comment:
Consider wrapping the `email[email.email] = true` statement in a block, to make it clearer you don't actually care about the return value, but rather a side effect.

Consider writing the whole thing in pure-functional style with `_.chain()`:

```javascript
return _.chain(Meteor.user.find({_id: {$in: SandstormDb.getUserIdentityIds(user)}}).fetch())
  .filter(identity => !services || services[identity.profile.service])
  .map(identity => SandstormDb.getVerifiedEmails(identity).map(email -> email.email))
  .flatten()
  .uniq()
  .value();
```